### ISSUE TITLE
Chapter3:Minist_Classify

### ISSUE THREAD

**Issue by anyuese** [ageron/handson-ml#344]
hello,I am a very new person in sklearn,I have a question while learing chapter3, the books writen this:
![image](https://user-images.githubusercontent.com/34855731/50810312-e9566280-1343-11e9-826e-be947935df52.png)

I know the goal is getting the descision scores,but why not use sgd_clf.decision_function()

**Comment by anyuese** [comment]
Thank you,master.I have another problem
![image](https://user-images.githubusercontent.com/34855731/50876429-f9d01100-1407-11e9-839b-f25617e26cad.png)
while using knn_fit and knn_predict,it just use a few time,but when using cross validation predicton then using f1_score() it cost me lots of time, I just think cv=3,the computation is around 3 times then just using knn_fit ,knn_predict. And while scoring f1,computation is fewer.But acutually,it's not,and I don't know why

**Comment by ageron** [comment]
Yes, KNN can be very slow.  Try running the code on 1/10th of the dataset to see if it runs smoothly.  Normally the cross val functions should be about close to 3 times slower when `cv=3`.

**Comment by qy-yang** [comment]
Hi @ageron,

Thanks for your explanation above. I have a question regarding the line `precisions, recalls, thresholds = precision_recall_curve(y_train_5, y_scores)`. Why the length of all the possible threshold is 59698 instead of 60000. As a naive way to think about this is that you can use every y_score as a possible threshold so that you can have 60000 sets of different prediction results. 

Thank you in advance.
Regards,
QY

**Comment by ageron** [comment]
Hi @qy-yang ,
Great question! I haven't checked, but I suppose these are all the *distinct* values.

**Comment by AlessandroMiola** [comment]
Hi @qy-yang, @ageron,
I had the same doubt expressed by @qy-yang. For me, given that the scores for this specific examples are all distinct (when I run `len(y_scores)` I get 60000), the point is the one specified [here](https://github.com/scikit-learn/scikit-learn/blob/84bd4e29680a9a95ca01143a6ba79a42bc6887ef/sklearn/metrics/_ranking.py#L819). Basically, the output is omitted for all thresholds that result in full recall, thus causing `thresholds` to be shorter than `y_scores`.
